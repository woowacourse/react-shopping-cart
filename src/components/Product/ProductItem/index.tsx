import * as S from './ProductItem.styles';
import Svg from 'components/@common/Svg';
import Counter from 'components/@common/Counter';
import { useProductSelect } from 'hooks/useProductSelect';
import { Product } from 'types';

interface ProductItemProps {
  product: Product;
}

const ProductItem = ({ product }: ProductItemProps) => {
  const { currentCartItem, remove, add, onSelectItem } =
    useProductSelect(product);

  return (
    <S.ItemWrapper>
      <S.ItemImage src={product.imageUrl} alt={product.name} />
      <S.ProductWrapper>
        <div>
          <S.ProductName>{product.name}</S.ProductName>
          <S.ProductPrice>
            {product.price.toLocaleString('KR')} Ïõê
          </S.ProductPrice>
        </div>
        {currentCartItem ? (
          <Counter
            count={currentCartItem.quantity}
            min={0}
            increment={add}
            decrement={remove}
          />
        ) : (
          <Svg type="cart-icon" width={25} height={22} onClick={onSelectItem} />
        )}
      </S.ProductWrapper>
    </S.ItemWrapper>
  );
};

export default ProductItem;
