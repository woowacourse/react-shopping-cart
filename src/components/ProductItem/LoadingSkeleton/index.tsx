import * as S from './LoadingSkeleton.styles';

const LoadingSkeleton = () => {
  return (
    <S.LoadingItemWrapper>
      <S.LoadingItemImage />
      <S.LoadingProductWrapper>
        <div>
          <S.LoadingProductName />
          <S.LoadingProductPrice />
        </div>
      </S.LoadingProductWrapper>
    </S.LoadingItemWrapper>
  );
};

export default LoadingSkeleton;
