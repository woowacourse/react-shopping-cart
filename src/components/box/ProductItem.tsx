import styled from '@emotion/styled';
import { CartIcon } from '../../assets';
import type { Product } from '../../types/types';
import { Text } from '../common/Text/Text';
import InputStepper from '../common/InputStepper/InputStepper';
import useCartList from '../../hooks/useCartList';

const ProductItem = ({ product }: { product: Product }) => {
  const { quantity, setQuantity } = useCartList(product);

  const handleOnClickToCartIcon = () => {
    setQuantity(1);
  };

  const handleSetQuantityOnInputStepper = (value: number) => {
    setQuantity(value);
  };

  return (
    <ProductWrapper>
      <ProductImage src={product.imageUrl} alt={product.name} />
      <ProductInfoWrapper>
        <ProductTextWrapper>
          <Text size="smallest" weight="light" color="#333333">
            {product.name}
          </Text>
          <Text size="small" weight="light" color="#333333" lineHeight="33px">
            {product.price} Ïõê
          </Text>
        </ProductTextWrapper>
        {quantity === 0 ? (
          <CartIcon
            width={25}
            height={22}
            fill="#AAAAAA"
            style={{ transform: 'scaleX(-1)', cursor: 'pointer' }}
            onClick={handleOnClickToCartIcon}
          />
        ) : (
          <InputStepper
            size="small"
            quantity={quantity}
            setQuantity={handleSetQuantityOnInputStepper}
          />
        )}
      </ProductInfoWrapper>
    </ProductWrapper>
  );
};

export default ProductItem;

const ProductWrapper = styled.div`
  display: flex;
  flex-direction: column;

  width: 282px;
`;
const ProductImage = styled.img`
  width: 100%;
  height: 282px;
`;

const ProductInfoWrapper = styled.div`
  display: flex;
  justify-content: space-between;

  width: 100%;
  padding: 12px;
`;

const ProductTextWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;
