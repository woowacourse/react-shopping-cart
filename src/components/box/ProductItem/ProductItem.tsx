import styled from '@emotion/styled';
import { CartIcon } from '../../../assets';
import type { Product } from '../../../types/types';
import { Text } from '../../common/Text/Text';
import InputStepper from '../../common/InputStepper/InputStepper';
import useCartList from '../../../hooks/useCartList';
import { formatPrice } from '../../../utils/formatPrice';

const ProductItem = ({ product }: { product: Product }) => {
  const { cartList, addProductToCartList, updateProductQuantity, removeProductInCartList } =
    useCartList();

  const productInCart = cartList?.find((cartItem) => cartItem.id === product.id);
  const quantityOfProductInCart = productInCart?.quantity ?? 0;

  const handleOnClickToCartIcon = () => {
    addProductToCartList(product.id);
  };

  const handleSetQuantityOnInputStepper = (quantity: number) => {
    updateProductQuantity(product.id, quantity);
  };

  const handleRemoveProductInCartList = () => {
    removeProductInCartList(product.id);
  };

  return (
    <ProductWrapper>
      <ProductImage src={product.imageUrl} alt={product.name} />
      <ProductInfoWrapper>
        <ProductTextWrapper>
          <Text size="smallest" weight="light" color="#333333">
            {product.name}
          </Text>
          <Text size="small" weight="light" color="#333333" lineHeight="33px">
            {formatPrice(product.price)} Ïõê
          </Text>
        </ProductTextWrapper>
        {quantityOfProductInCart === 0 ? (
          <CartIcon
            width={25}
            height={22}
            fill="#AAAAAA"
            style={{ transform: 'scaleX(-1)', cursor: 'pointer' }}
            onClick={handleOnClickToCartIcon}
          />
        ) : (
          <InputStepper
            size="small"
            quantity={quantityOfProductInCart}
            setQuantity={handleSetQuantityOnInputStepper}
            handleRemoveProductInCartList={handleRemoveProductInCartList}
          />
        )}
      </ProductInfoWrapper>
    </ProductWrapper>
  );
};

export default ProductItem;

const ProductWrapper = styled.div`
  display: flex;
  flex-direction: column;

  width: 282px;
`;

const ProductImage = styled.img`
  width: 100%;
  height: 282px;
`;

const ProductInfoWrapper = styled.div`
  display: flex;
  justify-content: space-between;

  width: 100%;
  padding: 12px;
`;

const ProductTextWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;
