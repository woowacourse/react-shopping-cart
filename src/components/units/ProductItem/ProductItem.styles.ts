import styled from '@emotion/styled';

const Root = styled.div`
  width: 282px;
  cursor: pointer;
`;

const ImageWrapper = styled.div`
  height: 282px;
  width: 282px;
`;

const Image = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

const Content = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px;
`;

const Detail = styled.div`
  letter-spacing: 0.5px;
  max-width: 200px;
`;

const Title = styled.div`
  font-size: 16px;
  line-height: 22px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const Price = styled.div`
  font-size: 20px;
  line-height: 27px;

  &::after {
    content: ' Ïõê';
  }
`;

const CartButton = styled.button`
  padding: 0.4em;
  border: none;
  background: none;
  cursor: pointer;

  &:hover {
    transform: scale(1.2);
  }
`;

export default { Root, ImageWrapper, Image, Content, Detail, Title, Price, CartButton };
