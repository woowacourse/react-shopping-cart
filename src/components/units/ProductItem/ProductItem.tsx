import React from 'react';
import { ReactComponent as CartIcon } from 'assets/images/cart.svg';
import * as T from 'types';
import noImageURL from 'assets/images/no_image.jpg';
import Styled from './ProductItem.styles';

type ProductItemProps = {
  product: T.Product;
  onClickCart: (product: T.Product) => void;
};

const ProductItem = (props: ProductItemProps) => {
  const { product, onClickCart } = props;
  const { name, image, price } = product;

  const handleClickCart = () => {
    onClickCart(product);
  };

  return (
    <Styled.Root>
      <Styled.ImageWrapper>
        <Styled.Image src={image || noImageURL} alt={name} />
      </Styled.ImageWrapper>
      <Styled.Content>
        <Styled.Detail>
          <Styled.Title>{name}</Styled.Title>
          <Styled.Price>{price.toLocaleString('ko-KR')}</Styled.Price>
        </Styled.Detail>
        <Styled.CartButton onClick={handleClickCart}>
          <CartIcon />
        </Styled.CartButton>
      </Styled.Content>
    </Styled.Root>
  );
};

export default ProductItem;
