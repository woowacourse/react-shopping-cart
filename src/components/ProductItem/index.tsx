import * as S from './ProductItem.styles';
import Svg from 'components/@common/Svg';
import Counter from '../Counter';
import useProductSelect from 'hooks/useProductSelect';
import { Product } from 'types';

interface ProductItemProps {
  product: Product;
}

const ProductItem = ({ product }: ProductItemProps) => {
  const { currentCartItem, remove, add, onSelectItem } =
    useProductSelect(product);

  const productSelect = currentCartItem ? (
    <Counter
      count={currentCartItem.quantity}
      increment={add}
      decrement={remove}
    />
  ) : (
    <Svg type="cart-icon" width={25} height={22} onClick={onSelectItem} />
  );

  return (
    <S.ItemWrapper>
      <S.ItemImage src={product.imageUrl} alt={product.name} />
      <S.ProductWrapper>
        <div>
          <S.ProductName>{product.name}</S.ProductName>
          <S.ProductPrice>
            {product.price.toLocaleString('KR')} Ïõê
          </S.ProductPrice>
        </div>
        {productSelect}
      </S.ProductWrapper>
    </S.ItemWrapper>
  );
};

export default ProductItem;
