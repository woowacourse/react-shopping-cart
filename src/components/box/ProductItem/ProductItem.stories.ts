import type { Meta, StoryObj } from '@storybook/react';
import ProductItem from './ProductItem';

const meta = {
  title: 'box/ProductItem',
  component: ProductItem,
  tags: ['autodocs'],
} satisfies Meta<typeof ProductItem>;

export default meta;

type Story = StoryObj<typeof meta>;

export const ProductContent: Story = {
  args: {
    product: {
      id: 1,
      name: 'PET보틀-정사각(370ml)',
      price: 41000,
      imageUrl:
        'https://s3-alpha-sig.figma.com/img/05ef/e578/d81445480aff1872344a6b1b35323488?Expires=1684713600&Signature=gvkVAezZeNJ6-pOzrsQBqsdfN9PGcEmV7HraRhfgfmXTZD0rCy-OEhFh-uP4y7AxUs5FC6ILRG-RNHslTX8x8q9HHcLyVrFq6I1uzwKaZyyWOCwAYgglD99UQlYPlVCU3E-8eIUfX~TRBM~dSPsojsKv-cNJU1gGPOE7G9C3Z1CKoDjIFZ8G2J4TG16VvSzil0RHsGGPi~06jOwn3Dtq~oBZQ1yhoRgfYDrGLlsexYockI38hGjBcBx1gJWg41~kBGNhueXs-tmuLfqRoPc0ReK~jWKLiGZi2JVPaIP-tPOfSAUEt99Po0KzVG-agBflDRhhpybhrR1bmozDvb~9eg__&Key-Pair-Id=APKAQ4GOSFWCVNEHN3O4',
    },
  },
};
