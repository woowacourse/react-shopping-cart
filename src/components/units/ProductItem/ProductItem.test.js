import React from 'react';
import { render } from '../../../test-utils';
import ProductItem from './ProductItem';

describe('ProductItem 테스트', () => {
  test('matches snapshot', () => {
    const product = {
      productId: 1,
      name: '재밌는 상품',
      price: 50000,
    };

    const utils = render(<ProductItem product={product} />);

    expect(utils.container).toMatchSnapshot();
  });
});
