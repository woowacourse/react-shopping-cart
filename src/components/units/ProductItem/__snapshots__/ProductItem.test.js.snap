// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem 테스트 matches snapshot 1`] = `
<div>
  <div
    class="css-1bqhmv4-Root etbhnjg7"
  >
    <div
      class="css-1iguuw0-ImageWrapper etbhnjg6"
    >
      <img
        alt="재밌는 상품"
        class="css-1azfbgt-Image etbhnjg5"
        src="no_image.jpg"
      />
    </div>
    <div
      class="css-2ln5zs-Content etbhnjg4"
    >
      <div
        class="css-wvtzlg-Detail etbhnjg3"
      >
        <div
          class="css-1wvdw29-Title etbhnjg2"
        >
          재밌는 상품
        </div>
        <div
          class="css-rhh1hx-Price etbhnjg1"
        >
          50,000
        </div>
      </div>
      <button
        class="css-qs5v00-CartButton etbhnjg0"
      >
        <svg>
          cart.svg
        </svg>
      </button>
    </div>
  </div>
</div>
`;
